Frameworks/Foundation/NSKVOSupport.mm

```diff
@@ -516,169 +515,301 @@ - (void)removeObserver:(id)observer forKeyPath:(NSString*)keyPath {
+                [change setObject:@(YES) forKey:NSKeyValueChangeNotificationIsPriorKey];
+                [observer observeValueForKeyPath:keypath ofObject:rootObject change:change context:context];
+                [change removeObjectForKey:NSKeyValueChangeNotificationIsPriorKey];
+            }
+
+            if (options & NSKeyValueObservingOptionNew) {
+                // Save old value in change dictionary for didChangeValueForKey:withSetMutation:usingObjects: to use for determining added
+                // objects
+                // Only needed if observer wants New value
+                change[_NSKeyValueChangeOldSetValue] = [[oldValues copy] autorelease];
+            }
+
+            keypathObserver.pendingChange = change;
+        });
     }
-    _dispatchDidChange(self, key);
 }
 
 /**
- @Status Stub
+ @Status Interoperable
 */
-- (void)willChangeValueForKey:(NSString*)key withSetMutation:(NSKeyValueSetMutationKind)mutationKind usingObjects:(NSSet*)options {
-    UNIMPLEMENTED();
-}
+- (void)didChangeValueForKey:(NSString*)key withSetMutation:(NSKeyValueSetMutationKind)mutationKind usingObjects:(NSSet*)objects {
+    if ([self observationInfo]) {
+        _dispatchDidChange(self, key, [mutationKind, objects](_NSKVOKeyObserver* keyObserver, NSMutableDictionary* keypathValueCache) {
+            _NSKVOKeypathObserver* keypathObserver = keyObserver.keypathObserver;
+            NSKeyValueObservingOptions options = keypathObserver.options;
+            id rootObject = keypathObserver.object;
+            id observer = keypathObserver.observer;
+            NSString* keypath = keypathObserver.keypath;
+            NSMutableDictionary* change = keypathObserver.pendingChange;
+            void* context = keypathObserver.context;
+            NSKeyValueChange changeKind = _changeFromSetMutationKind(mutationKind);
+
+            if ((options & NSKeyValueObservingOptionNew) && changeKind != NSKeyValueChangeRemoval) {
+                // For to-many mutations, we can't get the new values at indexes that have been deleted.
+                NSSet* newValue = keypathValueCache[keypath];
+                if (!newValue) {
+                    NSSet* oldValues = change[_NSKeyValueChangeOldSetValue];
+                    // The new value should only contain values which are added to the original dictionary
+                    // Objects added are in objects but not in oldValues
+
+                    newValue = [objects objectsPassingTest:^(id obj, BOOL* stop) {
+                        return [oldValues member:obj] ? NO : YES;
+                    }];
+                    keypathValueCache[keypath] = newValue;
+                }
 
-/**
- @Status Stub
-*/
-- (void)didChangeValueForKey:(NSString*)key withSetMutation:(NSKeyValueSetMutationKind)mutationKind usingObjects:(NSSet*)options {
-    UNIMPLEMENTED();
+                change[NSKeyValueChangeNewKey] = newValue;
+                change[_NSKeyValueChangeOldSetValue] = nil;
```

Comment:
please prefer `removeObjectForKey:` -- remember how painful it was to figure out that subscript assignment to nil deleted?